pkg/issuer/vault/issue.go

```diff
@@ -0,0 +1,200 @@
+package vault
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	vault "github.com/hashicorp/vault/api"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/util/errors"
+	"github.com/jetstack/cert-manager/pkg/util/kube"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	corelisters "k8s.io/client-go/listers/core/v1"
+)
+
+const (
+	errorGetCertKeyPair = "ErrGetCertKeyPair"
+	errorIssueCert      = "ErrIssueCert"
+
+	successCertIssued = "CertIssueSuccess"
+
+	messageErrorIssueCert = "Error issuing TLS certificate: "
+
+	messageCertIssued = "Certificate issued successfully"
+)
+
+const (
+	// certificateDuration of 90 days
+	certificateDuration = time.Hour * 24 * 90
```

Comment:
Should probably be configurable - perhaps as a field on the issuer?

Reply:
I agree. Initially, I was hoping for a shorter certificate duration but I hit the default renewBefore in pkg/controller/certificates/sync.go:

```golang
const renewBefore = time.Hour * 24 * 30
```

Maybe renewBefore and certificateDuration should be configurable? For Vault and CA issuers It could make sense I think. For ACME it is fixed to 90 days so It could be enforced at 30 days in that specific issuer.

Reply:
> For ACME it is fixed to 90 days so It could be enforced at 30 days in that specific issuer.

Careful, that's not true for ACME in general and might be over-fitting Let's Encrypt. The protocol allows flexibility in certificate lifetime.

Reply:
Oh ok sorry about that. Thanks for pointing it out!

Reply:
No need to be sorry :-)